Pseudobulk notebook (#6)
CellDiscoveryNetwork/workshops: navigating-the-single-cell-landscape-a-comprehensive-workshop-on-scrnaseq/day-1/4-psuedobulk.qmd

Author: change script name to 5-pseudobulk.qmd

> 

Author: Add this chunk to set options for all chunks below 
```suggestion
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, out.width = "100%", fig.align='center', 
                      message = FALSE, warning = FALSE, cache = FALSE)
options(width = 1200)
```
```

## Introduction

> Another way to process single-cell data is to pseudobulk the samples by aggregating the gene counts for cells belonging to each sample. This results in a single count value per gene per sample which is similar to a bulk matrix. Single-cell sparse matrices contain a lot of zeros which create a complicated distribution that makes it less sensitive to recognizing lowly expressed genes. Pseudobulking single-cell data allows scientists to benefit from the statistical rigor of existing bulk DE analysis on their single-cell data.

Author: rewrote it a bit!

Another way to process single-cell data is to pseudobulk the samples by aggregating the gene counts for cells belonging to each cell type by sample. This results in a single count value per gene per cell type per sample which is similar to a bulk matrix. Single-cell sparse matrices contain a lot of zeros which create a complicated distribution that makes it less sensitive to recognizing lowly expressed genes. Pseudobulking single-cell data allows us to benefit from the statistical rigor of existing bulk DE analysis.

### Useful Resources

-   [Bioconductor DESeq2 workflow](https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#interactions) ⭐️
-   [Bioinformagician DESeq2 workflow](https://www.youtube.com/watch?v=04gB2owLKus)
-   [Bioinformagician DESeq2 walk-through](https://www.youtube.com/watch?v=0b24mpzM_5M&t=1204s)

### Key Takeaways

> 1.  Aggregating gene counts at the sample level provides another way to determine differentially expressed genes in single-cell data.

Author: ```suggestion
1.  Aggregating gene counts by cell type at the sample level provides a way to determine differentially expressed genes in single-cell data.
```

2.  Pseudobulking single-cell data allows scientists to benefit from the statistical rigor of pre-existing bulk DE analysis on their single-cell data.

> 3.  Three

Author: ```suggestion
3.  Having a minimum of 3 replicates per condition is necessary to be able to obtain statistically significant results.
```

    BiocManager::install('EnhancedVolcano')
if (!requireNamespace("sparseMatrixStats", quietly = TRUE))
    BiocManager::install('sparseMatrixStats')

```

### Load Libraries

```{r}
suppressPackageStartupMessages({
  library(presto)
  library(DESeq2)
  library(tidyverse)
  library(Seurat)
  library(colorBlindness)
  library(RColorBrewer)
  library(org.Hs.eg.db)
  library(Matrix)
  library(tictoc)
  library(AnnotationDbi)
  library(DT)
  library(EnhancedVolcano)
  library(sparseMatrixStats)
  })

# Set seed to your favorite number for reproducibility
set.seed(21)
```

## Set working directory

Uncomment commands below to see what directory you are currently working in and set the correct one

```{r}
# getwd() 
# setwd('~/workshops/navigating-the-single-cell-landscape-a-comprehensive-workshop-on-scrnaseq/day-1/')
```

## Load Data

### Load Seurat object

```{r}
# 'se' or 'srobj' are common abbreviations for Seurat objects
srobj <- readRDS('../data/d8e35450-de43-451a-9979-276eac688bce.rds')

# 'genes' is a gene conversion table to convert ENSG to readable gene symbols
genes <- read_csv('../data/cov_flu_gene_names_table.csv') 
```

> 

Author: Explain the code chunk up here
```suggestion
Since we are modifying the names of the genes we need to recreate the Seurat object. In this process we need to ensure that mtx or contains the raw counts (integers) and not the normalized data. 
```

se
```

Set color palette

```{r}
pal <- paletteMartin
names(pal) <- sort(unique(se$Celltype))
```

## Analysis

### Seurat pre-processing

```{r, warnings = FALSE}
se <- se %>%
    NormalizeData(verbose = FALSE) %>%
    FindVariableFeatures(
        method = "vst",
        nfeatures = 3000,
        verbose = FALSE) %>%
    ScaleData(verbose = FALSE) %>%
    RunPCA(verbose = FALSE) %>%
    RunUMAP(dims = 1:30, verbose = FALSE, n.components=3L)

```

Visualize PCs

```{r, warnings = FALSE}
# Elbow plot of PCs
# NOTE: First 10-15 PCs hold most of the information
ElbowPlot(se, ndims = 30)
```

```{r, warnings = FALSE}
# Group cells by annotation
DimPlot(
    se,
    group.by = c("Celltype"),
    label = TRUE,
    cols = pal)
```

Modify strings with special characters to avoid complications later

```{r}
se@meta.data <- se@meta.data %>%
  mutate(
    Celltype = str_replace(Celltype, ", ", "_"),
    Celltype = str_replace(Celltype, "G\\-", "Gneg"),
    Celltype = str_replace(Celltype, "G\\+", "Gpos")
  )

```

Check to make sure strings were modified correctly before next steps

```{r}
unique(se@meta.data$Celltype)
```

### Aggregate Expression

> The AggregateExpression() function arranges the cell by gene matrix to group gene counts by the Sample ID and labelled Celltype of each cell. It outputs either a sparse matrix or a new seurat object if `return.Seurat = TRUE`. **Aggregation is the pseudobulking step where genes are summed across cells per celltype per sample.**

Author: References to packages and functions tend to go in between `...`
```suggestion
The `AggregateExpression()` function arranges the cell by gene matrix to group gene counts by the Sample ID and labelled Celltype of each cell. It outputs either a sparse matrix or a new `Seurat` object if `return.Seurat = TRUE`. **Aggregation is the pseudobulking step where genes are summed across cells per celltype per sample.**
```

```{r}
# Here we separate the dataset into two subsets
# CD8 and CD4 T cells
Idents(se) <- se@meta.data$Celltype
cd4sr <- subset(se, idents = c('CD4_EM-like', 'CD4_non-EM-like'))

> cd8sr <- subset(se, idents = c('CD8_non-EM-like', 'CD8_EM-like'))

Author: Remove since we're not gonna work with CD8s
```suggestion

```

# Filter for genes with 0 counts
cd4sr <- cd4sr[sparseMatrixStats::rowSums2(cd4sr@assays$RNA$data) > 0, ]

> cd8sr <- cd8sr[sparseMatrixStats::rowSums2(cd8sr@assays$RNA$data) > 0, ]

Author: Remove since we're not gonna work with CD8s
```suggestion

```

cd4_pseudobulk <- AggregateExpression(cd4sr, group.by = 'Sample ID')$RNA

> cd8_pseudobulk <- AggregateExpression(cd8sr, group.by = 'Sample ID')$RNA

Author: Remove since we're not gonna work with CD8s
```suggestion

```

dim(cd4_pseudobulk)

> dim(cd8_pseudobulk)

Author: Remove since we're not gonna work with CD8s
```suggestion

```

```

```{r}
# Save intermediate files because they might be useful later.
## TKTK : put at top or in "save" section
dir.create('pseudobulk_analysis')
dir.create('pseudobulk_analysis/data')

> ```

Author: Move this code to the top, after libraries

```{r}
## TKTK : put at top or in "save" section
list.files('pseudobulk_analysis/data')

> ```

Author: Move this code to the top, after libraries

## Run DESeq2

### Explanation

DESeq2 takes 3 main parameters.

1.  countData = A subset of the aggregated counts. Counts have to be whole numbers.

2.  colData = Sample metadata.

3.  design = Tells the DESeq function what comparisons to make.

> 

Author: ```suggestion

Prepare data to create a DESeq object
```

```{r}
# Here we run DESeq2, the classic differential expression package.

# 2 - Subset the sample metadata to match the sample x gene matrix.
sample_info <- strsplit(colnames(cd4_pseudobulk), " ")

lvls <- c("Normal", "Flu", "nCoV")
colData <- data.frame(
  sample = colnames(cd4_pseudobulk),
  condition = factor(sapply(sample_info, `[`, 1), levels = lvls)
)
DT::datatable(colData)

rownames(colData) <- colnames(cd4_pseudobulk)

# Note: If you encounter an error here, make sure dplyr is installed/running.

# Set "sample" variable to join the dataframes on
se@meta.data <- se@meta.data %>%
  dplyr::rename("sample" = "Sample ID")

# Join the metadata in se@meta.data to colData dataframe
colData <- colData %>%
  left_join(se@meta.data %>%
              dplyr::select('sample', 'Disease group') %>%
              distinct()
)
```

> 

Author: ```suggestion

Create `DEseq` object
```

```{r}
# Create the DESeq object, where tests can be run. 
## Check out the documentation online for more.
dds <- DESeqDataSetFromMatrix(countData = cd4_pseudobulk,
                                colData = colData,
                                design = ~ condition)
# this runs the test
dds <- DESeq(dds)
dds

## Plot dispersion estimates
# DESeq2 assumes that genes with similar expression levels have similar dispersion.
plotDispEsts(dds)
```

## Results

### What does this mean?

Results function examples:

-   res \<- results(dds)

-   res \<- results(dds, contrast = c("metadata_var", "opt1", "opt2"))

-   res \<- results(dds, name = "defined_condition")

Results output guide:

-   Positive log2fold change are up-regulated

-   Negative are down-regulated

-   Can set the p-value to a certain threshold

```{r}
results(dds, contrast = c("condition", "Flu", "Normal"))
```

> 

Author: ```suggestion

Extract results from the DESeq object
```

```{r}
# run 'resultsNames(dds)' to see "name" options
# current options: "condition_Flu_vs_Normal"  "condition_nCoV_vs_Normal"
# this pulls results from the DESeq object
res <- results(dds, name = "condition_Flu_vs_Normal")
res

# Summary function 
summary(res)
```

Save Results\
**NOTE: Save all data so if a collaborator asks what the comparison was, you have the header stored or rerun analysis.**

```{r}
# Uncomment below to create results folder
# dir.create('pseudobulk_analysis/results')

# Save as a R object
saveRDS(object = res, 
        file = 'pseudobulk_analysis/results/cd4_annot_pseudobulk_results.rds')

# Filtering and rearranging for plotting
restbl <- res %>% 
  data.frame %>% 
  rownames_to_column('gene')

# Write data frame results to CSV
write.csv(
  restbl,
  file='pseudobulk_analysis/results/cd4_annot_pseudobulk_results.csv'
  )
```

## Visualize Results

> Volcano plot is often the ultimate goal of differential expression analysis.

Author: ```suggestion
Volcano plot is one of the most utilized ways to visualize the results of differential expression analyses.
```

### From CSV object

```{r}
res <- read_csv('pseudobulk_analysis/results/cd4_annot_pseudobulk_results.csv')
```

```{r}
library(EnhancedVolcano)
EnhancedVolcano(res,
                lab = res$gene,
                x = 'log2FoldChange',
                y = 'padj',
                xlim = c(-8, 8),
                ylim = c(0, 5),
                title = 'Volcano Plot: CD4 T cells',
                subtitle = 'Normal < -  - - -  | -  - - -  > Flu',
                xlab = 'Log2 Fold Change',
                ylab = '-log10 FDR',
                pCutoff = 0.05,
                FCcutoff = 1.5,
                pointSize = c(ifelse(res$padj < 0.05 & abs(res$log2FoldChange) > 1.5, 1.5, 1)),
                labSize = 4.0,
                col = c('grey30', 'forestgreen', 'royalblue', 'red2'),
                colAlpha = 1,
                legendLabels=c('Not Significant', 'Log2FC', 'P-Value', 'Log2FC & P-Value'),
                legendPosition = 'right',
                legendLabSize = 14,
                drawConnectors = TRUE,
                widthConnectors = 0.5,
                colConnectors = 'grey30')

```

> TD: List notable genes, highlight where they are found, what they do, etc

Author: I think what I would highlight is that there is a general shift towards many genes being turned ON in disease context.

### From RDS object

```{r}
res <- readRDS('pseudobulk_analysis/results/cd4_annot_pseudobulk_results.rds')
```

> Volcano plot is often the ultimate goal of differential expression analysis.

Author: ```suggestion

```

```{r}
library(EnhancedVolcano)
EnhancedVolcano(res,
                lab = rownames(res),
                x = 'log2FoldChange',
                y = 'padj',
                xlim = c(-8, 8),
                ylim = c(0, 5),
                title = 'Volcano Plot: CD4 T cells',

>                 subtitle = 'Normal < -  - - -  | -  - - -  > Flu',

Author: ```suggestion
                subtitle = 'Normal < -  - - -  | - - -  -  > Flu',
```

                xlab = 'Log2 Fold Change',
                ylab = '-log10 FDR',
                pCutoff = 0.05,
                FCcutoff = 1.5,
                pointSize = c(ifelse(res$padj < 0.05 & abs(res$log2FoldChange) > 1.5, 1.5, 1)),
                labSize = 4.0,
                col = c('grey30', 'forestgreen', 'royalblue', 'red2'),
                colAlpha = 1,
                legendLabels=c('Not Significant', 'Log2FC', 'P-Value', 'Log2FC & P-Value'),
                legendPosition = 'right',
                legendLabSize = 14,
                drawConnectors = TRUE,
                widthConnectors = 0.5,
                colConnectors = 'grey30')

```

TD: List notable genes, highlight where they are found, what they do, etc

## Automate this for pre-annotated cell types

### Aggregate gene counts for each sample

> 

Author: Now we will showcase how to do this for all cell types in your dataset!

```{r}

> # TKTK: how do I filter for 0 gene counts of a seurat obj?
> # Filter genes with zero counts
> ## Start loop, then filter

Author: ```suggestion

```

> annot_pseudobulk <- AggregateExpression(

Author: ```suggestion
# Let's create pseudobulk profiles for all cell type by sample combinations
annot_pseudobulk <- AggregateExpression(
```

  se, # object name
  group.by = c('Sample ID','Celltype')
  )$RNA # desired assay
# NOTE: Can add "return.seurat = TRUE" and assign this to a new Seurat obj
```

> 

Author: ```suggestion

Prep the data to create a DESeq object
```

```{r}
# We have to rearrange the data somewhat to get it into the DESeq object. colData is the sample-level metadata

# Celltype variable is in question
sample_info <- strsplit(colnames(annot_pseudobulk), "_")

> # TKTK: figure out consistent factor order

Author: ```suggestion

```

# Prep variables
lvls <- c("Normal", "Flu", "nCoV")
ftr <- str_split(colnames(annot_pseudobulk), pattern = " ", simplify = TRUE)

colData <- data.frame(
  sample = colnames(annot_pseudobulk),
  condition = factor(ftr[, 1], levels = lvls),
  donor_id = sapply(sample_info, `[`, 1),
  CT = sapply(sample_info, `[`, 2)
)

# Ensure unique rownames in colData for DESeqDataSet
rownames(colData) <- colnames(annot_pseudobulk)

> colData <- colData %>%
>   left_join(se@meta.data %>%
>   dplyr::select(sample) %>%
>     distinct()
> ) %>%
>   column_to_rownames(var = 'sample')

Author: In this case I think its easier to read if everything in the left join is just one line
```suggestion
colData <- colData %>%
  left_join(
    se@meta.data %>% dplyr::select(sample) %>% distinct()
  ) %>%
  column_to_rownames(var = 'sample')
```

```

> 

Author: ```suggestion

Let's run the pseudobulk and DE analysis on all the celltypes
```

```{r}
# Use lapply function to loop through Celltypes

> lapply(unique(colData$CT), function(cluster) {

Author: Can you add comments on what each line does within the lapply loop? I know that we've gone through all of it with the CD4s but I think it will be helpful for people to have!

> lapply(unique(colData$CT), function(cluster) {

Author: Assign the output list to all_ct_ls
```suggestion
all_ct_ls <- lapply(unique(colData$CT), function(cluster) {
```

  split <- str_split_fixed(colnames(annot_pseudobulk),pattern='_',n=2)[,2]
  bulk_subset <- annot_pseudobulk[,split %in% cluster ]
  bulk_subset <- bulk_subset[sparseMatrixStats::rowSums2(bulk_subset@assays$RNA$data) > 0, ]
  data_subset <- colData %>% filter(CT %in% cluster) 
  data_subset$condition
  dds <- DESeqDataSetFromMatrix(countData = bulk_subset,
                                colData = data_subset,
                                design = ~ condition)
  
  cluster <- stringr::str_replace(cluster, " ", "_")
  c <- stringr::str_replace(cluster, "\\-", "_")
  dds <- DESeq(dds)
  # resultsNames(dds)
  res <- results(dds, name = "condition_Flu_vs_Normal")  

  r <- str_replace_all(c, '/', '.')
  saveRDS(
    res,
    glue::glue("pseudobulk_analysis/results/{r}_fluVnorm_annot_pseudobulk_results.rds"))
  
  restbl <- res %>%
  data.frame %>%
  rownames_to_column('gene')
  write_csv(
    restbl,

>     glue::glue("pseudobulk_analysis/results/{r}_fluVnorm_annot_pseudobulk_results.csv"))

Author: ```suggestion
    glue::glue("pseudobulk_analysis/results/{r}_fluVnorm_annot_pseudobulk_results.csv"))
    # Return restbl
    restbl
```

})

# TD: openxlsx package
# save each csv file individually, then loop through them again and put them into one excel file
```

Save to XLSX

```{r}

> # TKTK pick this up later
> # ls <- unique(colData$CT)
> # clust_xlsx <- lapply(names(ls), function(c){
> #   # print(c)
> #   # extract cluster
> #   print(c)
> # })

Author: Set a righ file path to save it
```suggestion
names(all_ct_ls) <- unique(colData$CT)
openxlsx::write.xlsx(
  all_ct_ls,
  file = "TKTK"
)
```

```

> 

Author: ```suggestion

Pull a list of all RDS files we saved within the loop
```

```{r}
res_ls <- list.files('pseudobulk_analysis/results/', pattern = "*.rds")
```

> 

Author: ```suggestion
Make a volcano plot for each cell type
```

    pdf(glue::glue('pseudobulk_analysis/figures/',strsplit(file,split='_')[[1]][1],'_volcano.pdf'),width=14,height=10)

    print(EnhancedVolcano(res_df,
                lab = res_df$gene,
                x = 'log2FoldChange',
                y = 'padj',
                xlim = c(-8, 8),
                ylim = c(0, 5),
                title = glue::glue('Volcano Plot: ',strsplit(file,split='_')[[1]][1]),
                subtitle = 'try again',
                xlab = 'Log2 Fold Change',
                ylab = '-log10 FDR',
                pCutoff = 0.05,
                FCcutoff = 1.5,
                pointSize = c(ifelse(res_df$padj < 0.05 & abs(res_df$log2FoldChange) > 1.5, 1.5, 1)),
                labSize = 4.0,
                col = c('grey30', 'forestgreen', 'royalblue', 'red2'),
                colAlpha = 1,
                legendLabels=c('Not Significant','Log2FC','P-Value','Log2FC & P-Value'),
                legendPosition = 'right',
                legendLabSize = 14,
                drawConnectors = TRUE,
                widthConnectors = 0.5,
                colConnectors = 'grey30'))
    dev.off()
})
    })
```

> 

Author: Code graveyard from here on out?

2.  load annotated, seurat object

-   use doublet determining info if you can

3.  preprocess data
4.  run seurat's normalization

`*The`echo: false\` option disables the printing of code (only output is displayed).\*

```{r}
sessionInfo()

> ```

Author: Keep this!
